skimage/measure/_moments.py

```diff
@@ -1,11 +1,137 @@
+
+    Returns
+    -------
+    m : (``order + 1``, ``order + 1``) array
+        Raw image moments.
+
+    References
+    ----------
+    .. [1] Johannes Kilian. Simple Image Analysis By Moments. Durham
+           University, version 0.2, Durham, 2001.
+
+    Examples
+    --------
+    >>> contour = np.array([[r, c] for r in range(13, 17) \
+                  for c in range(13, 17)], dtype=np.double)
+    >>> M = moments_contour(contour)
+    >>> cr = M[1, 0] / M[0, 0]
+    >>> cc = M[0, 1] / M[0, 0]
+    >>> cr, cc
+    (14.5, 14.5)
+    """
+    return moments_contour_central(contour, 0, order=order)
+
+
+def moments_contour_central(contour, center=None, order=3):
+    """Calculate all raw image moments up to a certain order.
+
+    The following properties can be calculated from raw image moments:
+     * Area as: ``M[0, 0]``.
+     * Centroid as: {``M[1, 0] / M[0, 0]``, ``M[0, 1] / M[0, 0]``}.
+
+    Note that raw moments are neither translation, scale nor rotation
+    invariant.
+
+    Parameters
+    ----------
+    contour : (N, D) double or uint8 array
+        Array of N points that describe an image of D dimensionality in
+        Euclidian space.
+    center : tuple of float, optional
+        Coordinates of the image centroid. This will be computed if it
+        is not provided.
+    order : int, optional
+        Maximum order of moments. Default is 3.
+
+    Returns
+    -------
+    m : (``order + 1``, ``order + 1``) array
+        Raw image moments.
+
+    References
+    ----------
+    .. [1] Johannes Kilian. Simple Image Analysis By Moments. Durham
+           University, version 0.2, Durham, 2001.
+
+    Examples
+    --------
+    >>> contour = np.array([[r, c] for r in range(13, 17) \
+                  for c in range(13, 17)], dtype=np.double)
+    >>> M = moments_contour(contour)
+    >>> cr = M[1, 0] / M[0, 0]
+    >>> cc = M[0, 1] / M[0, 0]
+    >>> moments_contour_central(contour, (cr, cc))
+    array([[ 16.,   0.,  20.,   0.],
+           [  0.,   0.,   0.,   0.],
+           [ 20.,   0.,  25.,   0.],
+           [  0.,   0.,   0.,   0.]])
+    """
+    if isinstance(contour, tuple):
+        # Represent as array and rotate to standardize orientation.
+        contour = np.rollaxis(np.asarray(contour), axis=1)
+    assert_nD(contour, 2)
+    ndim = contour.shape[1]
+    if center is None:
+        M = moments_central(contour, center=0, order=order)
+        center = M[tuple(np.eye(ndim, dtype=int))]
+    else:
+        # TODO: Handle broadcasting of `center` argument.
+        if center is int:
+            center = (center,) * ndim
+
+    calc = contour.astype(float)
+    calc -= center
+
+    # generate all possible exponents for each axis in the given set of points
+    # produces a matrix of shape (N, D, order + 1)
+    calc = calc[..., np.newaxis] ** np.arange(order + 1)
+
+    # FIXME: Find a better way to do this than DIY `fold` function.
+    def _expand(current, mat):
```

Comment:
Yeah, as indicated by your comment, I don't understand what you're doing here...!

Reply:
Think of it like this:
```python
In [1]: import numpy as np

In [2]: def expand(mat): _expand(mat[0], mat[1:])

In [3]: a = np.arange(1, 9).reshape(2, 4)

In [4]: a
Out[4]: 
array([[1, 2, 3, 4],
       [5, 6, 7, 8]])

In [5]: expand(a)
Out[5]: 
array([[ 5,  6,  7,  8],
       [10, 12, 14, 16],
       [15, 18, 21, 24],
       [20, 24, 28, 32]])

In [6]: b = np.arange(1, 13).reshape(3, 4)

In [6]: b
Out[6]: 
array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [7]: (b[0][..., np.newaxis] * b[1])[..., np.newaxis] * b[2]
Out[7]: 
array([[[ 45,  50,  55,  60],
        [ 54,  60,  66,  72],
        [ 63,  70,  77,  84],
        [ 72,  80,  88,  96]],

       [[ 90, 100, 110, 120],
        [108, 120, 132, 144],
        [126, 140, 154, 168],
        [144, 160, 176, 192]],

       [[135, 150, 165, 180],
        [162, 180, 198, 216],
        [189, 210, 231, 252],
        [216, 240, 264, 288]],

       [[180, 200, 220, 240],
        [216, 240, 264, 288],
        [252, 280, 308, 336],
        [288, 320, 352, 384]]])

```

i.e., given an `(M, N)` matrix, `expand` will generate an `(N, ) * M` matrix of all possible products of the elements of each row multiplied by the elements of the other rows. I am not well-versed in the ways of numpy so if something like this already exists, please let me know!

Reply:
ah, so it's kind of like reducing the column vectors with `np.outer`...? =) Is M the dimensions or the number of points here???

Reply:
@jni In this application of `expand`, `M` is the number of dimensions of the coordinates and `N` is the order of the moments to calculate.

Similar, but not quite the same. While `np.outer` will perform what we wish for two vectors, folding over our array for cases where M > 2 (our n-dimensional applications), it will produce an `(N**M)xN` array, which is the improper shape. However, the numbers are all correct – meaning that we can apply `reshape( (N, ) * M )` to our result. If you think that this might be faster, we can investigate it, but a few quick tests that I ran suggest otherwise.